src/writer.rs

```diff
@@ -119,168 +52,185 @@ pub fn write_nifti<A, S, D>(
-    let resolution: usize = header.dim[1..(header.dim[0] + 1) as usize]
-        .iter().map(|d| *d as usize).product();
-    let nb_bytes = resolution * header.bitpix as usize / 8;
-    let dim0_size = data.shape()[0];
+    let mut write_slice = |data: Array<A, D::Smaller>| -> Result<()> {
+        let len = data.len();
+        let arr_data = data.into_shape(len).unwrap();
+        let slice = arr_data.as_slice().unwrap();
+        writer.write_all(guarded_transmute_to_bytes_pod_many(slice))?;
+        Ok(())
+    };
 
     // `1.0x + 0.0` would give the same results, but we avoid a lot of divisions
-    let slope = if header.scl_slope == 0.0 { 1.0 } else { header.scl_slope };
+    let slope = if header.scl_slope == 0.0 {
+        1.0
+    } else {
+        header.scl_slope
+    };
     if slope != 1.0 || header.scl_inter != 0.0 {
         let slope = A::from_f32(slope).unwrap();
         let inter = A::from_f32(header.scl_inter).unwrap();
         for arr_data in data.axis_iter(Axis(0)) {
-            let arr_data = arr_data.sub(inter).div(slope);
-            let arr_data = arr_data.into_shape(arr_len).unwrap();
-            let arr_data = arr_data.as_slice().unwrap();
-
-            let buffer = to_bytes(&arr_data, nb_bytes / dim0_size);
-            writer.write_all(&buffer).unwrap();
+            write_slice(arr_data.sub(inter).div(slope))?;
         }
     } else {
         for arr_data in data.axis_iter(Axis(0)) {
-            // We need to own the data because of the into_shape() for `C` ordering.
-            let arr_data = arr_data.to_owned();
-            let arr_data = arr_data.into_shape(arr_len).unwrap();
-            let arr_data = arr_data.as_slice().unwrap();
-
-            let buffer = to_bytes(&arr_data, nb_bytes / dim0_size);
-            writer.write_all(&buffer).unwrap();
+            // We need to own the data because of the into_shape() for `c` ordering.
+            write_slice(arr_data.to_owned())?;
         }
     }
-}
-
-fn to_bytes<T>(data: &[T], size: usize) -> &[u8] {
-    unsafe { from_raw_parts::<u8>(data.as_ptr() as *const u8, size) }
+    Ok(())
 }
 
 #[cfg(test)]
 pub mod tests {
     extern crate tempfile;
 
     use super::*;
+    use std::ops::{Add, Mul};
+    use std::path::PathBuf;
 
-    use ndarray::{Array, Array2, Ix2, IxDyn, ShapeBuilder};
     use self::tempfile::tempdir;
+    use ndarray::{Array, Array2, Ix2, IxDyn, ShapeBuilder};
 
-    use {InMemNiftiObject, IntoNdArray, header::MAGIC_CODE_NIP1, object::NiftiObject};
+    use {header::MAGIC_CODE_NIP1, object::NiftiObject, InMemNiftiObject, IntoNdArray};
 
-    fn get_random_path(ext: &str) -> String {
+    fn get_random_path(ext: &str) -> PathBuf {
         let dir = tempdir().unwrap();
         let path = if ext == "" {
             dir.into_path()
         } else {
             dir.into_path().join(ext)
         };
-        path.to_str().unwrap().to_string()
+        PathBuf::from(path.to_str().unwrap())
```

Comment:
Isn't `path` already a `PathBuf`? It should be, according to [`into_path()`](https://docs.rs/tempfile/3.0.4/tempfile/struct.TempDir.html#method.into_path).